index.js

```diff
@@ -111,8 +111,8 @@ const oauthPlugin = fp(function (fastify, options, next) {
 
   if (startRedirectPath) {
     fastify.get(startRedirectPath, startRedirectHandler)
-    fastify.decorate('getAccessTokenFromAuthorizationCodeFlow', getAccessTokenFromAuthorizationCodeFlow)
-    fastify.decorate('getNewAccessTokenUsingRefreshToken', getNewAccessTokenUsingRefreshToken)
+    Object.assign(oauth2, { getAccessTokenFromAuthorizationCodeFlow })
+    Object.assign(oauth2, { getNewAccessTokenUsingRefreshToken })
```

Comment:
I'm thinking if we could implement this feature like the [`namespace`](https://github.com/fastify/fastify-redis#registering-multiple-redis-client-instances) does, in order to don't change the `oauth2` object.

I would like to do some tests and think a little bit more because it could be good to add this feature as semver minor.

Reply:
What if
```
fastify.decorate('github', {
  oauth2: oauth2,
  getAccessTokenFromAuthorizationCodeFlow,
  getNewAccessTokenUsingRefreshToken,
})
```
???

In this way we can register different providers allowing the developer to choose the name (it is like namespace)

Reply:
Simple and clean 👍

Reply:
I understand that with the initial solution provided in this PR, we are going to have breaking changes and we would like to avoid that.

But I don't quite understand how would your config look like in the case of registering multiple providers with this suggestion?

Currently, we are registering this plugin multiple times with diff configs for diff providers, like this:

```javascript
fastify.register(require('fastify-oauth2'), {
  name: 'facebook',
  credentials: {
    client: {
      id: process.env.FACEBOOK_ID || null,
      secret: process.env.FACEBOOK_SECRET || null
    },
    auth: oauthPlugin.FACEBOOK_CONFIGURATION
  },
  startRedirectPath: '/oauth/login/facebook',
  callbackUri: '/@/oauth/callback/facebook'
})

fastify.register(require('fastify-oauth2'), {
  name: 'github',
  credentials: {
    client: {
      id: process.env.GITHUB_ID || null,
      secret: process.env.GITHUB_SECRET || null
    },
    auth: oauthPlugin.GITHUB_CONFIGURATION
  },
  startRedirectPath: '/oauth/login/github',
  callbackUri: '/@/oauth/callback/github'
})
```

And using it like this:

```javascript
// some route handler
async oauth (request, reply) {
    const type = request.query.type
    const oauthConfig = this[type]
    if (!oauthConfig) {
      throw new TypeError(`Unsupported OAuth type: ${type}`)
    }
    const access = await oauthConfig.getAccessTokenFromAuthorizationCodeFlow(request)
    // get data using this token...
  }
```


I assume what you would like to achieve is to register it once with a single config that contains sub-configs for diff providers?

Also I'm not clear how is this not causing a breaking change, since in order to access e.g. `getAccessTokenFromAuthorizationCodeFlow`, we would still need to do like this:

`this.github.getAccessTokenFromAuthorizationCodeFlow`

which looks like a breaking change again.

Hence, I would appreciate if you can provide examples:
1. How we register the plugin with multiple providers
2. How it's fastify is decorated with this helper methods
3. How we access those helper methods in the callback from the OAuth provider.

Thanks.

Reply:
Forget about the version, as suggested also by @allevo we will have a major 👍

(side note: to avoid major I thought to add a flag that registers the `getAccessTokenFromAuthorizationCodeFlow` in the `name`, otherwise old code, but it is not worth)

> How we register the plugin with multiple providers

Your example is right: one `.register` for every provider needed

> How fastify is decorated with these helper methods

Only one decorator:

```
fastify.decorate(name, {
  oauth2: oauth2,
  getAccessTokenFromAuthorizationCodeFlow,
  getNewAccessTokenUsingRefreshToken,
})
```

> How we access those helper methods in the callback from the OAuth provider.

The `name` became mandatory to access all the methods and the `simple-oauth2 ` client:

```
// assuming we have registered 2 providers with name: spotifyProvider and facebookProvider:

const token = await this.spotifyProvider.getAccessTokenFromAuthorizationCodeFlow(request)
...
const simpleClient = this.spotifyProvider.oauth2
..
const token = await this.spotifyProvider.getAccessTokenFromAuthorizationCodeFlow(request)
```

Reply:
Ok got it.

In that case, this is totally fine with me.

I will push a new commit with this approach then:
```javascript
fastify.decorate(name, {
  oauth2: oauth2,
  getAccessTokenFromAuthorizationCodeFlow,
  getNewAccessTokenUsingRefreshToken,
})
```

Reply:
Alright, change pushed and tests are passing.

Pls review and let me know if anything else is needed.